cmd/validator/validate.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	ctrlApp "github.com/atlassian/ctrl/app"
+	"github.com/atlassian/voyager/pkg/apis/composition/v1"
+	"github.com/atlassian/voyager/pkg/composition"
+	"github.com/atlassian/voyager/pkg/options"
+	"github.com/atlassian/voyager/pkg/util/crash"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/yaml"
+)
+
+func main() {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	defer crash.LogPanicAsJSON()
+	ctrlApp.CancelOnInterrupt(ctx, cancelFunc)
+
+	err := runWithContext(ctx)
+
+	if err != nil && errors.Cause(err) != context.Canceled {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func runWithContext(context context.Context) error {
+	sd, err := LoadSD("cmd/validator/testdata/sd.yaml")
+	if err != nil {
+		return err
+	}
+
+	// Composition
+	clusterLocation, err := LoadLocation("cmd/validator/testdata/location.yaml")
```

Comment:
Not sure if I understood your idea correctly, but Kubernetes now supports "dry run" mode which we can use instead of having a separate binary?
See https://github.com/kubernetes/kubernetes/issues/11488
it's a server-side feature that is enabled via passing a flag inside create/update requests, e.g. https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L498-L504

Need to confirm that this flag is respected by `apiextensions-apiserver` (CRD).

Reply:
Intention is to separate the code that does the useful work from the kubernetes code around it so we can, say, run validation without round tripping through 4 objects or 4 separate dry run calls.

Reply:
Sorry but I don't see this validator going through 4 objects. At the moment it does more or less the same as what the replication validation webhook does, i.e. taking SD and trying to transform it into "formation objects"? Do you plan to extend this validator to do the whole chain of SD->LD->State->Bundle conversion? That would also require location-specific config, service metadata, deployment artifacts and other things?

Reply:
Need to rewrite all of the controllers to make it possible. 
> Do you plan to extend this validator to do the whole chain of SD->LD->State->Bundle conversion 

yes

> location-specific config, service metadata, deployment artifacts and other things?

yes